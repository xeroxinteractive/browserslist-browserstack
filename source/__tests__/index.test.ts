import * as moduleUnderTest from '../index';
import mockFetch from '../__mocks__/node-fetch';

describe('getCapabilities', () => {
  describe('browserslist options', () => {
    test('single browser query', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'IE 10'
          }
        })
      ).resolves.toMatchSnapshot();
    });

    test('multiple browser query', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: ['IE 10', 'Edge 18']
          }
        })
      ).resolves.toMatchSnapshot();
    });

    test('last 2 versions', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'last 2 versions'
          }
        })
      ).resolves.toMatchSnapshot();
    });

    test('invalid query', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'invalid'
          }
        })
      ).rejects.toEqual(
        new moduleUnderTest.BrowsersListError(
          'Unknown browser query `invalid`. Maybe you are using old Browserslist or made typo in query.'
        )
      );
    });

    test('invalid credentials', async () => {
      mockFetch.mockImplementationOnce((...args) =>
        require.requireActual('node-fetch').default(...args)
      );
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'invalid'
          }
        })
      ).rejects.toEqual(new moduleUnderTest.ResponseError('Unauthorized', 401));
    });
  });
  describe('OS Filtering', () => {
    test('os + os version include filters', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'IE 10'
          },
          operatingSystems: {
            include: [moduleUnderTest.OperatingSystemFilter.WINDOWS]
          },
          operatingSystemVersion: {
            include: [
              moduleUnderTest.WindowsOperatingSystemVersionFilter.SEVEN,
              moduleUnderTest.WindowsOperatingSystemVersionFilter.TEN,
              moduleUnderTest.WindowsOperatingSystemVersionFilter.XP
            ]
          }
        })
      ).resolves.toMatchSnapshot();
    });

    test('os + os version exclude filters', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'Firefox 42'
          },
          operatingSystems: {
            exclude: [moduleUnderTest.OperatingSystemFilter.OSX]
          },
          operatingSystemVersion: {
            exclude: [
              moduleUnderTest.WindowsOperatingSystemVersionFilter.EIGHT_ONE,
              moduleUnderTest.WindowsOperatingSystemVersionFilter.EIGHT
            ]
          }
        })
      ).resolves.toMatchSnapshot();
    });

    test('os + os version include + exclude filters', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: 'Firefox 42'
          },
          operatingSystems: {
            include: [moduleUnderTest.OperatingSystemFilter.OSX],
            exclude: [moduleUnderTest.OperatingSystemFilter.WINDOWS]
          },
          operatingSystemVersion: {
            exclude: [
              moduleUnderTest.OSXOperatingSystemVersionFilter.LION,
              moduleUnderTest.OSXOperatingSystemVersionFilter.HIGH_SIERRA,
              moduleUnderTest.WindowsOperatingSystemVersionFilter.EIGHT
            ]
          }
        })
      ).resolves.toMatchSnapshot();
    });
  });

  describe('browser filtering', () => {
    test('browser include filter', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: ['Chrome 72', 'Opera 12', 'Edge 18', 'IE 10']
          },
          browsers: {
            include: [
              moduleUnderTest.BrowserFilter.EDGE,
              moduleUnderTest.BrowserFilter.IE
            ]
          }
        })
      ).resolves.toMatchSnapshot();
    });
    test('browser exclude filter', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: ['Chrome 72', 'Opera 12', 'Edge 18', 'IE 10']
          },
          browsers: {
            exclude: [
              moduleUnderTest.BrowserFilter.OPERA,
              moduleUnderTest.BrowserFilter.CHROME
            ]
          }
        })
      ).resolves.toMatchSnapshot();
    });
    test('browser include + exclude filter', async () => {
      await expect(
        moduleUnderTest.default({
          browserslist: {
            queries: ['Chrome 72', 'Opera 12', 'Edge 18', 'IE 10']
          },
          browsers: {
            include: [moduleUnderTest.BrowserFilter.EDGE],
            exclude: [
              moduleUnderTest.BrowserFilter.OPERA,
              moduleUnderTest.BrowserFilter.CHROME
            ]
          }
        })
      ).resolves.toMatchSnapshot();
    });

    test('formatting for selenium', async () => {
      const capabilities = await moduleUnderTest.default({
        browserslist: {
          queries: 'IE 10'
        },
        formatForSelenium: false
      });
      for (const capability of capabilities) {
        expect(capability).not.toHaveProperty('browserName');
        expect(capability).not.toHaveProperty('browserVersion');
      }
    });
  });
});
