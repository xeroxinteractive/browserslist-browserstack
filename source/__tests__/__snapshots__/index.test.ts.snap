// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCapabilities OS Filtering os + os version exclude filters 1`] = `
Array [
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "Windows",
    "os_version": "XP",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities OS Filtering os + os version include + exclude filters 1`] = `
Array [
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Snow Leopard",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mountain Lion",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mavericks",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Yosemite",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "El Capitan",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mojave",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "42.0",
    "browser_version": "42.0",
    "device": null,
    "os": "OS X",
    "os_version": "Catalina",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities OS Filtering os + os version include filters 1`] = `
Array [
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities browser filtering browser exclude filter 1`] = `
Array [
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "edge",
    "browserName": "edge",
    "browserVersion": "18.0",
    "browser_version": "18.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities browser filtering browser include + exclude filter 1`] = `
Array [
  Object {
    "browser": "edge",
    "browserName": "edge",
    "browserVersion": "18.0",
    "browser_version": "18.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities browser filtering browser include filter 1`] = `
Array [
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "edge",
    "browserName": "edge",
    "browserVersion": "18.0",
    "browser_version": "18.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities browserslist options last 2 versions 1`] = `
Array [
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "11.0",
    "browser_version": "11.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "Windows",
    "os_version": "8.1",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "Windows",
    "os_version": "8.1",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "11.0",
    "browser_version": "11.0",
    "device": null,
    "os": "Windows",
    "os_version": "8.1",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "Windows",
    "os_version": "8.1",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "Windows",
    "os_version": "8.1",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "11.0",
    "browser_version": "11.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "edge",
    "browserName": "edge",
    "browserVersion": "17.0",
    "browser_version": "17.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "edge",
    "browserName": "edge",
    "browserVersion": "18.0",
    "browser_version": "18.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mavericks",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mavericks",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "OS X",
    "os_version": "Yosemite",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "OS X",
    "os_version": "Yosemite",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "Yosemite",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "Yosemite",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "OS X",
    "os_version": "El Capitan",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "OS X",
    "os_version": "El Capitan",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "El Capitan",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "El Capitan",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "OS X",
    "os_version": "Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "OS X",
    "os_version": "Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "OS X",
    "os_version": "High Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "OS X",
    "os_version": "High Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "High Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "High Sierra",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mojave",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mojave",
    "real_mobile": null,
  },
  Object {
    "browser": "safari",
    "browserName": "safari",
    "browserVersion": "12.1",
    "browser_version": "12.1",
    "device": null,
    "os": "OS X",
    "os_version": "Mojave",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mojave",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "Mojave",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "72.0",
    "browser_version": "72.0",
    "device": null,
    "os": "OS X",
    "os_version": "Catalina",
    "real_mobile": null,
  },
  Object {
    "browser": "chrome",
    "browserName": "chrome",
    "browserVersion": "73.0",
    "browser_version": "73.0",
    "device": null,
    "os": "OS X",
    "os_version": "Catalina",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "65.0",
    "browser_version": "65.0",
    "device": null,
    "os": "OS X",
    "os_version": "Catalina",
    "real_mobile": null,
  },
  Object {
    "browser": "firefox",
    "browserName": "firefox",
    "browserVersion": "66.0",
    "browser_version": "66.0",
    "device": null,
    "os": "OS X",
    "os_version": "Catalina",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities browserslist options multiple browser query 1`] = `
Array [
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
  Object {
    "browser": "edge",
    "browserName": "edge",
    "browserVersion": "18.0",
    "browser_version": "18.0",
    "device": null,
    "os": "Windows",
    "os_version": "10",
    "real_mobile": null,
  },
]
`;

exports[`getCapabilities browserslist options single browser query 1`] = `
Array [
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "7",
    "real_mobile": null,
  },
  Object {
    "browser": "ie",
    "browserName": "ie",
    "browserVersion": "10.0",
    "browser_version": "10.0",
    "device": null,
    "os": "Windows",
    "os_version": "8",
    "real_mobile": null,
  },
]
`;
